### PR TITLE
53 add dao property to task

### LINKED ISSUE #53: Add DAO property to Task
Currently, in the DAO pallet we have a collection of tasks that belong to a certain DAO. The task is agnostic to which organization it belongs. This design can be improved by adding an additional property to the TASK struct where we can state the organization to which the task belongs. 

The task may not belong to any organization, so it is an optional field.

- [x] Refactor code to add Dao property to task
- [x] Add Tests
- [x] Remove unnecessary functionality in DAO pallet where tasks are added to DAO
- [x] Verify DAO identifier (via loose coupling) during task creation/updates

### PR DESCRIPTION
**Pull request by evilrobot-01**
- [x] Refactor code to add Dao property to task
- [x] Add Tests
- [x] Remove unnecessary functionality in DAO pallet where tasks are added to DAO
- [x] Verify DAO identifier (via loose coupling) during task creation/updates

NOTE: I used the ports & adapters pattern for the loose coupling, whereby the `Organization` trait defined within the `Task` pallet simply defines a function which determines whether a particular organization identifier exists. This trait is then used as a trait bound on the `Organization` type in the pallet config, with the implementation in two places:

1. `mock.rs:Test` - implements the trait for the Test mock runtime, forwarding the call to the `Dao` pallet
2. `runtime/src/lib.rs:Runtime` - as above, implements the trait for the `Runtime`, forwarding the call to the `Dao` pallet

Finally, the pallet dependencies are only added as `dev-dependencies` to the Task pallet, meaning that they are only pulled in for configuration of the mock runtime during tests.

### CODE REVIEW

**Comment by evilrobot-01** [comment]
The tests in the `Task` pallet had to be updated to use public keys for the `AccountId` due to the dependency on the `Dao` pallet when constructing the mock runtime.

My current understanding is that the mock runtime should simulate the actual running node as much as possible but in a testable way, so could perhaps be thought of as integration tests to some degree. Might be good to have additional unit tests for non-dispatchable functionality, where practical.

**Comment by evilrobot-01** [comment]
I wonder if the type alias at https://github.com/UniversalDot/universal-dot-node/blob/8f47ed3496129c7dcf31f6b90155eb7bc6961aa4/pallets/dao/src/lib.rs#L144 should rather be `DaoIdOf<T>`?

This was only introduced to differentiate between the previous Task identifier hash and the Dao identifier hash.

File: pallets/dao/src/lib.rs [hunk 1]
```diff
@@ -155,6 +141,7 @@ pub mod pallet {
 
 	// Account used in Dao Struct
 	type AccountOf<T> = <T as frame_system::Config>::AccountId;
+	type DaoOf<T> = <T as frame_system::Config>::Hash;
```
Comment (f-gate) [inline pallets/dao/src/lib.rs]: Agreed that should be DaoIdOf<T>

**Review by f-gate — CHANGES_REQUESTED** [review]
Looks good for me, thanks for fixing my grammatical mistakes and whitespace discrepancies. I'll make sure they're proper next time.

File: pallets/task/src/mock.rs [hunk 2]
```diff
@@ -24,10 +24,12 @@ frame_support::construct_runtime!(
 		UncheckedExtrinsic = UncheckedExtrinsic,
 	{
 		System: frame_system::{Pallet, Call, Config, Storage, Event<T>},
-		Task: pallet_task::{Pallet, Call, Storage, Event<T>},
 		Balances: pallet_balances::{Pallet, Call, Storage, Config<T>, Event<T>},
+		Did: pallet_did::{Pallet, Call, Storage, Event<T>},
```
Comment (stojanov-igor) [inline pallets/task/src/mock.rs:28]: Why we need the DID and DAO pallet in the task mock?

File: pallets/task/src/tests.rs [hunk 3]
```diff
@@ -136,178 +170,174 @@ fn increase_task_count_when_creating_task(){
...
+		assert_ok!(Task::create_task(Origin::signed(*ALICE), title(), spec(), BUDGET, get_deadline(), attachments(), keywords(), None));
+		assert_ok!(Task::create_task(Origin::signed(*ALICE), title(), spec2(), BUDGET2, get_deadline(), attachments(), keywords(), None));
 
 		// Assert that count is incremented to 2 after task creation
 		assert_eq!(Task::task_count(), 2);
 	});
 }
 
 #[test]
-fn cant_own_more_tax_than_max_tasks(){
```
Comment (stojanov-igor) [inline pallets/task/src/tests.rs:153]: Tax :D

File: pallets/task/src/tests.rs [hunk 4]
```diff
@@ -316,531 +346,496 @@ fn task_can_be_updated_only_by_one_who_created_it(){
...
 
 		// Ensure new task is assigned to new current_owner (user 1)
-		let hash = Task::tasks_owned(1)[0];
-		let task = Task::tasks(hash).expect("should found the task");
-		assert_eq!(task.current_owner, 1);
-		assert_eq!(Task::tasks_owned(1).len(), 1);
+		let task_id = Task::tasks_owned(*ALICE)[0];
+		let task = Task::tasks(task_id).expect("should found the task");
+		assert_eq!(task.current_owner, *ALICE);
+		assert_eq!(Task::tasks_owned(*ALICE).len(), 1);
```
Comment (stojanov-igor) [inline pallets/task/src/tests.rs:420]: I think we are using the aliases in DAO. Its much more readable than using numbers as accounts. Its a good improvement.

File: pallets/task/src/lib.rs [hunk 5]
```diff
@@ -151,6 +157,8 @@ pub mod pallet {
 		pub created_at: <T as frame_system::Config>::BlockNumber,
 		pub updated_at:<T as frame_system::Config>::BlockNumber,
 		pub completed_at: <T as frame_system::Config>::BlockNumber,
+		/// The organization to which the task belongs.
+		pub dao: Option<OrganizationIdOf<T>>
```
Comment (stojanov-igor) [inline pallets/task/src/lib.rs]: organization would be a more appropriate name. Your suggestion yesterday was correct. Can we rename dao to organization.

**Review by stojanov-igor — APPROVED** [review]
Few minor suggestions for changes.

File: pallets/task/src/mock.rs (hunk 2 above)
Comment (evilrobot-01) [reply pallets/task/src/mock.rs:28]: The `Dao` pallet is required to satisfy the `OrganizationId` verification during task creation/updates. The tests make use of organisation verification.

The `Did` pallet needs to be pulled in due to the tight coupling of `Dao` pallet to the `Did` pallet.

File: pallets/task/src/tests.rs (hunk 3 above)
Comment (evilrobot-01) [reply pallets/task/src/tests.rs:153]: Should I change it to 'cant_owe_more_tax'? :p

File: pallets/task/src/tests.rs (hunk 4 above)
Comment (evilrobot-01) [reply pallets/task/src/tests.rs:420]: Yeah, I copied it from the Dao pallet.

File: pallets/dao/src/lib.rs (hunk 1 above)
Comment (evilrobot-01) [reply pallets/dao/src/lib.rs]: Implemented, thanks!

File: pallets/task/src/lib.rs (hunk 5 above)
Comment (evilrobot-01) [reply pallets/task/src/lib.rs]: Done, will be in next commit.

**Review by f-gate — APPROVED** [review]

**Pull request merged**